MapboxVisionTests/Mocks/MockVideoSource.swift

```diff
@@ -1,10 +1,93 @@
 import Foundation
+
 @testable import MapboxVision
 
-class MockVideoSource: VideoSource {
-    var isExternal: Bool = true
+final class MockVideoSource: ObservableVideoSource {
+    private enum Constants {
+        static let fps: Int32 = 30
+        static let frameDelay = 1.0 / Double(fps)
+        static let frameWidth = 960
+        static let frameHeight = 540
+        static let bitsInByte = 8
+    }
+
+    private let processingQueue = DispatchQueue(label: "com.mapbox.RecordCoordinatorTest.VideoSource")
+    private lazy var timer: DispatchSourceTimer = {
+        let timer = DispatchSource.makeTimerSource()
+        timer.schedule(deadline: .now() + Constants.frameDelay, repeating: Constants.frameDelay)
+        timer.setEventHandler { [weak self] in
+            self?.emitNewBuffer()
+        }
+        return timer
+    }()
+    private var framesAlreadyWritten: Int64 = 0
+
+    override init() {
+        super.init()
+    }
+
+    func start() {
+        timer.resume()
+    }
+
+    func stop() {
+        timer.suspend()
+    }
+
+    private func emitNewBuffer() {
+        var buffer: CMSampleBuffer?
+
+        let pixelBuffer = makePixelBuffer()!
+        let duration = CMTime(value: 1, timescale: Constants.fps)
+        let presentationTimeStamp = CMTime(value: framesAlreadyWritten, timescale: Constants.fps)
+        var sampleTimingInfo = CMSampleTimingInfo(duration: duration, presentationTimeStamp: presentationTimeStamp, decodeTimeStamp: CMTime.invalid)
+        CMSampleBufferCreateReadyWithImageBuffer(allocator: kCFAllocatorDefault,
+                                                 imageBuffer: pixelBuffer,
+                                                 formatDescription: make(from: pixelBuffer)!,
+                                                 sampleTiming: &sampleTimingInfo,
+                                                 sampleBufferOut: &buffer)
+
+        framesAlreadyWritten += 1
+        let sample = VideoSample(buffer: buffer!, format: Image.Format.BGRA)
+
+        self.notify { observer in
+            observer.videoSource(self, didOutput: sample)
+        }
+    }
+
+    func makePixelBuffer() -> CVPixelBuffer? {
+        let attrs = [
+            kCVPixelBufferCGImageCompatibilityKey: kCFBooleanTrue,
+            kCVPixelBufferCGBitmapContextCompatibilityKey: kCFBooleanTrue,
+            kCVPixelBufferOpenGLESCompatibilityKey: kCFBooleanTrue,
+            kCVPixelBufferOpenGLESTextureCacheCompatibilityKey: kCFBooleanTrue,
```

Comment:
It doesn't seem like we need those keys.

Reply:
removed all but metal compatibility in order to have an ability to use the class with our renderer